source/slang/slang-emit-spirv.cpp

```diff
@@ -0,0 +1,1141 @@
+// instructions.
+//
+// We will start by forward-declaring the type we will
+// use to represent instructions:
+//
+struct SpvInst;
+
+// Next, we will define a base type that can serve as a parent
+// to SPIR-V instructions. Both the logical sections defined
+// earlier and instructions such as functions will be used
+// as parents.
+
+    /// Base type for SPIR-V instructions and logical sections of a module
+    ///
+    /// Holds and supports appending to a list of child instructions.
+struct SpvInstParent
+{
+public:
+        /// Add an instruction to the end of the list of children
+    void addInst(SpvInst* inst);
+
+        /// Dump all children, recursively, to a flattened list of SPIR-V words
+    void dumpTo(List<SpvWord>& ioWords);
+
+private:
+        /// The first child, if any.
+    SpvInst* m_firstChild = nullptr;
+
+        /// A pointer to the null pointer at the end of the linked list.
+        ///
+        /// If the list of children is empty this points to `m_firstChild`,
+        /// while if it is non-empty it points to the `nextSibling` field
+        /// of the last instruction.
+        ///
+    SpvInst** m_link = &m_firstChild;
+};
+
+// A SPIR-V instruction is then (in the general case) a potential
+// parent to other instructions.
+
+    /// A type to represent a SPIR-V instruction to be emitted.
+    ///
+    /// This type alows the instruction to be built up across
+    /// multiple steps in a mutable fashion.
+    ///
+struct SpvInst : SpvInstParent
+{
+    // [2.3: Physical Layout of a SPIR-V Module and Instruction]
+    //
+    // > Each instruction is a stream of words
+    //
+    // > Opcode: The 16 high-order bits are the WordCount of the instruction.
+    // >         The 16 low-order bits are the opcode enumerant.
+    //
+    // We will store the "opcode enumerant" directly in our
+    // intermediate structure, and compute the word count on
+    // the fly when writing an instruction to an output buffer.
+
+        /// The SPIR-V opcode for the instruction
+    SpvOp opcode;
+
+    // [2.3: Physical Layout of a SPIR-V Module and Instruction]
+    //
+    // > Optional instruction type <id> (presence determined by opcode)
+    // > Optional instruction Result <id> (presence determined by opcode)
+    // > Operand 1 (if needed)
+    // > Operand 2 (if needed)
+    // > ...
+    //
+    // We represent the remaining words of the instruction (after
+    // the opcode word) as an undifferentiated array. Any code
+    // that encodes an instruction is responsible for knowing the
+    // opcode-specific data that is required.
+    //
+    // Our code does not need to process instruction operands after
+    // they have been written into a `SpvInst`. If we ever had
+    // cases where we needed to do post-processing, then we would
+    // need to store a more refined representation here.
+
+        /// The additional words of the instruction after the opcode
+    List<SpvWord> operandWords;
```

Comment:
Yes, the allocation policies here could use a lot of work. The main reason I didn't try to do anything clever is simply that I'm not as familiar with those utilities (because you wrote most of them). If you can give me a quick pointer to an example I could use as a starting point for how to do arena-based allocation the Right Way using your utilities, that would be helpful.